File: templates/python/src/mcp_server/http.py
"""Streamable HTTP transport, stateless mode.

`MCPServer.streamable_http_app()` returns a Starlette app whose lifespan runs
the SDK's session manager. We wrap it in a small ASGI middleware that enforces
a bearer token on the MCP path and add a /healthz route.

Stateful variant (needed for server-initiated notifications): pass
`stateless_http=False`; the SDK then issues an Mcp-Session-Id header and keeps
per-session state in memory, so route each session to one replica.
"""

from __future__ import annotations

import hmac
import logging
from typing import Any

from mcp.server.transport_security import TransportSecuritySettings
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.types import ASGIApp, Receive, Scope, Send

from . import SERVER_NAME, SERVER_VERSION
from .api_client import ApiClient
from .config import Config
from .server import create_server

log = logging.getLogger(__name__)


class BearerAuthMiddleware:
    def __init__(self, app: ASGIApp, token: str, protected_path: str) -> None:
        self.app = app
        self.token = token.encode()
        self.protected_path = protected_path.rstrip("/")

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http" or not scope["path"].rstrip("/").startswith(self.protected_path):
            await self.app(scope, receive, send)
            return
        header = dict(scope.get("headers", [])).get(b"authorization", b"")
        presented = header[7:] if header.startswith(b"Bearer ") else b""
        if not hmac.compare_digest(presented, self.token):
            response = JSONResponse(
                {"error": "unauthorized"}, status_code=401, headers={"WWW-Authenticate": 'Bearer realm="mcp"'}
            )
            await response(scope, receive, send)
            return
        await self.app(scope, receive, send)


def create_http_app(config: Config, api: ApiClient) -> ASGIApp:
    mcp = create_server(api)

    @mcp.custom_route("/healthz", methods=["GET"])
    async def healthz(_: Request) -> JSONResponse:
        return JSONResponse({"status": "ok", "name": SERVER_NAME, "version": SERVER_VERSION})

    security: TransportSecuritySettings | None = None
    loopback = config.http_host in ("127.0.0.1", "localhost", "::1")
    if config.allowed_hosts:
        security = TransportSecuritySettings(
            enable_dns_rebinding_protection=True, allowed_hosts=config.allowed_hosts
        )
    elif not loopback:
        log.warning(
            "binding to %s without MCP_ALLOWED_HOSTS; DNS rebinding protection is off", config.http_host
        )
        security = TransportSecuritySettings(enable_dns_rebinding_protection=False)

    app: Any = mcp.streamable_http_app(
        streamable_http_path=config.http_path,
        stateless_http=True,
        json_response=False,
        transport_security=security,
        host=config.http_host,
    )

    if config.auth_token:
        app = BearerAuthMiddleware(app, config.auth_token, config.http_path)
    else:
        log.warning("MCP_AUTH_TOKEN is not set; the MCP endpoint is unauthenticated")
    return app


__all__ = ["BearerAuthMiddleware", "Starlette", "create_http_app"]

File: templates/typescript/src/http.ts
/**
 * Streamable HTTP transport, stateless mode.
 *
 * Every POST /mcp builds a fresh server and transport, handles the request,
 * and tears both down when the response closes. Nothing is kept between
 * requests, so any number of replicas can sit behind a load balancer.
 *
 * Stateful variant (needed for server-initiated notifications): create the
 * transport with `sessionIdGenerator: () => randomUUID()`, keep a
 * Map<sessionId, transport>, look up the transport by the `mcp-session-id`
 * header on later requests, and also route GET (SSE stream) and DELETE
 * (session end) to it. See the SDK's simpleStreamableHttp.ts example.
 */
import { createMcpExpressApp } from "@modelcontextprotocol/sdk/server/express.js";
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import type { Express, NextFunction, Request, Response } from "express";
import { timingSafeEqual } from "node:crypto";
import type { ApiClient } from "./api-client.js";
import type { Config } from "./config.js";
import { log } from "./log.js";
import { createServer, SERVER_NAME, SERVER_VERSION } from "./server.js";

function bearerAuth(expected: string) {
  const expectedBuf = Buffer.from(expected);
  return (req: Request, res: Response, next: NextFunction): void => {
    const header = req.headers.authorization ?? "";
    const token = header.startsWith("Bearer ") ? header.slice(7) : "";
    const tokenBuf = Buffer.from(token);
    const ok = tokenBuf.length === expectedBuf.length && timingSafeEqual(tokenBuf, expectedBuf);
    if (!ok) {
      res.setHeader("WWW-Authenticate", 'Bearer realm="mcp"');
      res.status(401).json({ error: "unauthorized" });
      return;
    }
    next();
  };
}

export function createHttpApp(config: Config, api: ApiClient): Express {
  const app = createMcpExpressApp({
    host: config.httpHost,
    allowedHosts: config.allowedHosts.length > 0 ? config.allowedHosts : undefined,
  });

  app.get("/healthz", (_req, res) => {
    res.json({ status: "ok", name: SERVER_NAME, version: SERVER_VERSION });
  });

  if (config.authToken) {
    app.use(config.httpPath, bearerAuth(config.authToken));
  } else {
    log.warn("MCP_AUTH_TOKEN is not set; the MCP endpoint is unauthenticated");
  }

  app.post(config.httpPath, async (req, res) => {
    const server = createServer(api);
    const transport = new StreamableHTTPServerTransport({ sessionIdGenerator: undefined });
    res.on("close", () => {
      void transport.close();
      void server.close();
    });
    try {
      await server.connect(transport);
      await transport.handleRequest(req, res, req.body);
    } catch (err) {
      log.error("request failed", { error: String(err) });
      if (!res.headersSent) {
        res.status(500).json({
          jsonrpc: "2.0",
          error: { code: -32603, message: "Internal server error" },
          id: null,
        });
      }
    }
  });

  // Stateless servers have no session stream to resume or delete.
  const methodNotAllowed = (_req: Request, res: Response) => {
    res.status(405).json({
      jsonrpc: "2.0",
      error: { code: -32000, message: "Method not allowed" },
      id: null,
    });
  };
  app.get(config.httpPath, methodNotAllowed);
  app.delete(config.httpPath, methodNotAllowed);

  return app;
}
